### PR TITLE
feat(client): expose a public API to query chunks to specific data replicas

### PR DESCRIPTION
**Pull request by bochaco**
- Exposing also an `sn_api` public API to fetch a file from a specified set of data replicas indexes and a `SafeUrl`.
- Adding `--replicas` arg to CLI `dog` command which allows the user to perform a check on several data replicas for the content being targeted by specifying their indexes.
- It currently supports querying data replicas only for `Files`, either with a `File`'s `XorUrl` or with any `SafeUrl` which resolves to a File.
- Non-pretty printing is also supported for the data replicas report, i.e. when passing `--output <format>` arg.

Example output from CLI querying `my-file.txt` linked from a FilesContainer to replicas with indexes 0, 2, 3, and 10 (index out of range of actual replicas), when some chunks were not found in any replica:
```
$ safe dog safe://hyryyrytx4imnzxfrwydbuy7fmtru56q5mjhnw1ig4ro6o1ea7u43adwh6wnra/my-file.txt -r0 -r2 -r3 -r10

== URL resolution step 1 ==
Resolved from: safe://hyryyrytx4imnzxfrwydbuy7fmtru56q5mjhnw1ig4ro6o1ea7u43adwh6wnra
= FilesContainer =
XOR-URL: safe://hyryyrytx4imnzxfrwydbuy7fmtru56q5mjhnw1ig4ro6o1ea7u43adwh6wnra
Version: hxhtb38edw54getr4onmjrrndmp1ry6ohne5m9f71iofe474h91hy
Type tag: 1100
XOR name: 0x2fd5562bbca4a0061983a55c493df9db5a782a4aa6d121e84918ecf59c0e9cf5
Native data type: Register
Native data XOR-URL: safe://hyryyyytx4imnzxfrwydbuy7fmtru56q5mjhnw1ig4ro6o1ea7u43adwh6wnra

== URL resolution step 2 ==
Resolved from: safe://hyryyyyxwih81u5xiwhtk9hfo47qyrnjpusryucx1barp7nkrfn7uqdkkfr
= File =
XOR-URL: safe://hyryyyyxwih81u5xiwhtk9hfo47qyrnjpusryucx1barp7nkrfn7uqdkkfr
XOR name: 0xf4af0f29edf5a722aff0b0d75c02092d9d8809b1f20e08de894428bb370d4a29
Native data type: PublicFile
Media type: Unknown

== Checking data replicas of resolved content ==
XOR-URL: safe://hyryyyyxwih81u5xiwhtk9hfo47qyrnjpusryucx1barp7nkrfn7uqdkkfr
Replicas indexes queried: [0, 2, 3, 10]
Content composed of 4 chunk/s:
= Chunk at XOR name 0xf4af0f29edf5a722aff0b0d75c02092d9d8809b1f20e08de894428bb370d4a29 =
Replica #0: Ok!
Replica #2: Ok!
Replica #3: Ok!
Replica #10: Error received from the network: InsufficientNodeCount { prefix: Prefix(), expected: 11, found: 4 }

= Chunk at XOR name 0x7086f5f1570884db5d062cb4a0d1334d88db83a5dda8263aa74cc2bc81bbc4d9 =
Replica #0: Ok!
Replica #2: Error received from the network: DataNotFound(Bytes(ChunkAddress(7086f5(01110000)..)))
Replica #3: Ok!
Replica #10: Error received from the network: InsufficientNodeCount { prefix: Prefix(), expected: 11, found: 4 }

= Chunk at XOR name 0xb1551a74e9a3404f0afa086c511f11e0d797eaa80143c5206dcce0875b2af52b =
Replica #0: Ok!
Replica #2: Ok!
Replica #3: Ok!
Replica #10: Error received from the network: InsufficientNodeCount { prefix: Prefix(), expected: 11, found: 4 }

= Chunk at XOR name 0x53127c013fdd0fa1d2b385d55d30843e8b0fbd0752fe43db4a93d41dd7af9aeb =
Replica #0: Error received from the network: DataNotFound(Bytes(ChunkAddress(53127c(01010011)..)))
Replica #2: Ok!
Replica #3: Ok!
Replica #10: Error received from the network: InsufficientNodeCount { prefix: Prefix(), expected: 11, found: 4 }
```

### CODE REVIEW

**Comment by bochaco** [comment]
bors try

File: sn_cli/src/subcommands/dog.rs [hunk 1]
```diff
@@ -148,14 +154,47 @@ pub async fn dog_commander(cmd: DogCommands, output_fmt: OutputFmt, safe: &Safe)
...
+                    println!("XOR-URL: {xorurl}");
+                    println!("Type tag: {type_tag}");
                     println!("XOR name: 0x{}", xorname_to_hex(xorname));
                     println!("Native data type: Register");
                 }
             }
+
+            // If this is the last resolution step, and a set of replicas indexes was provided,
+            // then query all data replicas and print out a report.
+            if !replicas_indexes.is_empty() && i == num_of_resolutions - 1 {
```
Comment (jacderida) [inline sn_cli/src/subcommands/dog.rs:168]: If it wasn't too difficult it would be nice if we could get a CLI test case for the new argument.

Comment (bochaco) [reply sn_cli/src/subcommands/dog.rs:168]: yeah, sounds good, I'll try to add one

Comment (bochaco) [reply sn_cli/src/subcommands/dog.rs:168]: ...it seems I'll need to add support for non-pretty printing to be able to do so, thus I'll add support for that and a couple of tests if possible

File: sn_cli/src/subcommands/dog.rs [hunk 2]
```diff
@@ -24,12 +24,16 @@ use tracing::debug;
 pub struct DogCommands {
     /// The safe:// location to inspect
     location: Option<String>,
+    /// Query all the data replicas matching the given indexes to check they hold a
+    /// copy of the content. E.g. -r0 -r2 will query replicas at index 0 and 2.
```
Comment (joshuef) [inline sn_cli/src/subcommands/dog.rs:28]: I think we need to elaborate a bit on what a data index is here. 

A data index is when we sort of list of nodes in a section by a data name, the node at that index in the list. It a measure of "closeness to the data" in xorspace. Only the first `data_copy_count` nodes should hold this data. More may well hold it too.

Or something... feels a bit verbose there

**Review by joshuef — APPROVED** [review]
LGTM in general. Minor comment re:docs, and with @jacderida 'ss test suggestion I think we're good to get this in 👍

**Comment by bochaco** [comment]
I've pushed a commit taking care of the comments.

**Review by jacderida — APPROVED** [review]

**Comment by bochaco** [comment]
bors merge

**Comment by bochaco** [comment]
bors try

**Comment by bochaco** [comment]
bors try-

**Comment by bochaco** [comment]
bors merge

**Pull request merged**